src/components/Login/Signup.js

```diff
@@ -0,0 +1,159 @@
+import React, { useEffect, useReducer, useRef, useState } from "react";
+import Button from "../UI/Button";
+import classes from "./Login.module.css";
+import LoginInput from "./LoginInputs";
+import LoginModal from "../UI/LoginModal";
+import { useAuth } from "../contexts/AuthContext";
+import { Link, useNavigate } from "react-router-dom";
+
+const emailReducer = (state, action) => {
+  if (action.type === "USER_EMAIL_INPUT") {
+    return { value: action.val, isValid: !!action.val.match(emailFormat) };
+  }
+
+  if (action.type === "USER_EMAIL_BLUR") {
+    return { value: state.value, isValid: !!state.value.match(emailFormat) };
+  }
+
+  return { value: "", isValid: null };
+};
+
+const passwordReducer = (state, action) => {
+  if (action.type === "USER_PASSSWORD_INPUT") {
+    return { value: action.val, isValid: action.val.trim().length > 6 };
+  }
+
+  if (action.type === "USER_PASSWORD_BLUR") {
+    return { value: state.value, isValid: state.value.trim().length > 6 };
+  }
+  return { value: "", isValid: null };
+};
+
+const emailFormat = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+
+export function Signup(props) {
+  const emailRef = useRef();
+  const passwordRef = useRef();
+  const passwordConfirmRef = useRef();
+  const { signup, currentUser } = useAuth();
+  const [error, setError] = useState("");
+  const [Loading, setLoading] = useState(false);
+  const [formIsValid, setFormIsValid] = useState(false);
+  const navigate = useNavigate();
+
+  const [emailState, dispatchEmail] = useReducer(emailReducer, {
+    value: "",
+    isValid: null,
+  });
+
+  const [passwordState, dispatchPassword] = useReducer(passwordReducer, {
+    value: "",
+    isValid: null,
+  });
+
+  const { isValid: emailIsValid } = emailState;
+  const { isValid: passwordIsValid } = passwordState;
+
+  useEffect(() => {
+    const timeIdentifier = setTimeout(() => {
+      console.log("Checking form validity");
+      setFormIsValid(emailIsValid && passwordIsValid);
+    }, 500);
+    return () => {
+      console.log("CLEANUP");
+      clearTimeout(timeIdentifier);
+    };

```

Comment:
This is silly. Never use timeouts... A more effective way to check if the form is valid is by putting this function into an `onChange` input event.